src/java/org/apache/cassandra/cql3/statements/SelectStatement.java

```diff
@@ -446,14 +455,14 @@ private void warn(String msg)
     {
         int userLimit = getLimit(options);
         int userPerPartitionLimit = getPerPartitionLimit(options);
-        int pageSize = options.getPageSize();
+        PageSize pageSize = options.getPageSize();
 
         Selectors selectors = selection.newSelectors(options);
         ReadQuery query = getQuery(state, options, selectors.getColumnFilter(), nowInSec, userLimit, userPerPartitionLimit, pageSize);
 
         try (ReadExecutionController executionController = query.executionController())
         {
-            if (aggregationSpec == null && (pageSize <= 0 || (query.limits().count() <= pageSize)))
+            if (aggregationSpec == null && !pageSize.isCompleted(query.limits().count(), PageSize.PageUnit.ROWS))
```

Comment:
Should this also include `&& !pageSize.isCompleted(query.limits().bytes(), PageSize.PageUnit.BYTES)`?

Reply:
Actually, that previous change was wrong. So basically the check:
```
        if (aggregationSpec == null && !pageSize.isCompleted(query.limits().count(), PageSize.PageUnit.ROWS))
            return execute(query, options, queryState, selectors, nowInSec, userLimit, queryStartNanoTime);
```
should verify we paging should be applied at all. What we want to do there is to verify whether the requested page size is greater than the user provided limits. Say if the user specifies `LIMIT 50` and the page size in rows is 100, then paging is unnecessary. However if the page size is provided in bytes, we cannot say anything about that. I've extracted that condition into a separate method and added a comment.